File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        result = []

        # Use inorder traversal for BST and push node's value into result as it goes
        # Note that result will be sorted naturally because of the nature of BST
        def dfs(node: Optional[TreeNode]) -> None:
            if node is None:
                return
            dfs(node.left)
            result.append(node.val)
            dfs(node.right)

        dfs(root)

        # Return kth smallest element
        return result[k - 1]

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    bool isValidBST(TreeNode* root) {
        auto dfs = [](this auto&& self, TreeNode* node, long long left, long long right) -> bool {
            // Return true if current node is empty because we have nothing to check
            if (node == nullptr) {
                return true;
            }
            // Return false if current node is not valid for any of left/right bounds
            if (!(node->val > left && node->val < right)) {
                return false;
            }
            return self(node->left, left, node->val) && self(node->right, node->val, right);
        };

        // Note that TreeNode's value is int so we're using `long long` to avoid the boundary
        // problem For example, if we use INT_MIN and if that node's value is INT_MIN, the bool
        // check will not return what we expected. Thus, we use a wider range and guarantee to have
        // 8 bytes in all platforms, which is long long.
        return dfs(root, LLONG_MIN, LLONG_MAX);
    }
};
